### ISSUE TITLE
quantity_simplify() broke solve for quantities

### ISSUE THREAD

**Issue by schymans** [sympy/sympy#14595]
Sorry for the lengthy example, but on the current master, this gives an error:
```python
from sympy import Eq, solve
from sympy.physics.units import Quantity, energy, length, time, mass, joule, kelvin, kilogram, length, meter, second

H_l = Quantity('H_l')
H_l.set_scale_factor(joule/meter**2/second)
H_l.set_dimension(energy/length**2/time)

H_u = Quantity('H_u')
H_u.set_scale_factor(joule/meter**2/second)
H_u.set_dimension(energy/length**2/time)

Q_d = Quantity('Q_d')
Q_d.set_scale_factor(joule/meter**2/second)
Q_d.set_dimension(energy/length**2/time)

R_li = Quantity('R_li')
R_li.set_scale_factor(joule/meter**2/second)
R_li.set_dimension(energy/length**2/time)

R_lo = Quantity('R_lo')
R_lo.set_scale_factor(joule/meter**2/second)
R_lo.set_dimension(energy/length**2/time)

R_ui = Quantity('R_ui')
R_ui.set_scale_factor(joule/meter**2/second)
R_ui.set_dimension(energy/length**2/time)

R_uo = Quantity('R_uo')
R_uo.set_scale_factor(joule/meter**2/second)
R_uo.set_dimension(energy/length**2/time)

solve(Eq(-H_u - Q_d + R_ui - R_uo, -H_l + Q_d + R_li - R_lo), Q_d)
```
gives:

> ...
> ~/Programs/sympy/sympy/physics/units/util.py in quantity_simplify(expr)
>     160             if len(v) > 1 else v[0][0]**v[0][1]
>     161         for quantity, (k, v) in zip(ref_quantities, quantity_pow_by_dim.items())]
> --> 162     return coeff*Mul.fromiter(other_pow)*Mul.fromiter(new_quantities)
> 
> ~/Programs/sympy/sympy/core/containers.py in __mul__(self, other)
>      87             n = as_int(other)
>      88         except ValueError:
> ---> 89             raise TypeError("Can't multiply sequence by non-integer of type '%s'" % type(other))
>      90         return self.func(*(self.args*n))
>      91 
> 
> TypeError: Can't multiply sequence by non-integer of type '<class 'sympy.core.numbers.Half'>'

If I follow a suggestion by @jirikuncar and replace https://github.com/sympy/sympy/blob/70ff5651edf91985ebf523a970fb88ec07240799/sympy/physics/units/util.py#L151-L162 by `return expr`, I get the correct result: 

> [H_l/2 - H_u/2 - R_li/2 + R_lo/2 + R_ui/2 - R_uo/2]

**Comment by ishanaj** [comment]
I am getting the correct answer:
```
>>> from sympy import Eq, solve
>>> from sympy.physics.units import Quantity, energy, length, time, mass, joule, kelvin, kilogram, length, meter, second
>>> H_l = Quantity('H_l', energy/length**2/time, joule/meter**2/second)
>>> H_u = Quantity('H_u', energy/length**2/time, joule/meter**2/second)
>>> Q_d = Quantity('Q_d', energy/length**2/time, joule/meter**2/second)
>>> R_li = Quantity('R_li', energy/length**2/time, joule/meter**2/second)
>>> R_lo = Quantity('R_lo', energy/length**2/time, joule/meter**2/second)
>>> R_ui = Quantity('R_ui', energy/length**2/time, joule/meter**2/second)
>>> R_uo = Quantity('R_uo',energy/length**2/time, joule/meter**2/second)

>>> solve(Eq(-H_u - Q_d + R_ui - R_uo, -H_l + Q_d + R_li - R_lo), Q_d)
[H_l/2 - H_u/2 - R_li/2 + R_lo/2 + R_ui/2 - R_uo/2]
```

Also, it shows a `TypeError` when I declare the `Quantity` as @schymans has described.
```
>>> H_l = Quantity('H_l')
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-882d6e5da79e> in <module>()
----> 1 H_l = Quantity('H_l')

TypeError: __new__() missing 1 required positional argument: 'dimension'
``` 
Why this error?

**Comment by smichr** [comment]
In master you now get [0] with the default `simplify=True` because the units are all "roses by another name" and simplify as such.

When simplifying `inch - foot` the canonical answer is 
```console
>>> quantity_simplify(inch - foot)
-11*foot/12
```
In your case, all units have the same scale factor and the canonical answer is 0. cf #14374
```console
>>> eq = Eq(-H_u - Q_d + R_ui - R_uo, -H_l + Q_d + R_li - R_lo)
>>> solve(eq, Q_d, simplify=False)
[H_l/2 - H_u/2 - R_li/2 + R_lo/2 + R_ui/2 - R_uo/2]
>>> simplify(_[0])
0
```

**Comment by smichr** [comment]
Should simplify be set to false when solving for a Quantity?

**Comment by schymans** [comment]
Thanks for looking into this!
The example I put forward is not very good. It is motivated by this package: https://github.com/environmentalscience/essm, where we originally used `Quantity()` to represent physical constants and variables. We ran into all sorts of problems. Now we base our `Variable` class on `Symbol` and consider quantities just as units. Works like a charm. Perhaps this could be merged into sympy at some point, but for the moment, I keep working with essm. I think everything would be a lot clearer if we used the current `Quantity()` for units and a separate `Variable` for physical constants and variables as above. Then there would be no point in defining different quantities with the same dimension and scale factor and the above simplification problem would not occur. I am closing this for now as it confuses the purpose of `Quantity()`.
